src/ui/widget.rs

```diff
@@ -0,0 +1,206 @@
+//! Reusable UI widgets.
+
+use bevy::{
+    ecs::system::{EntityCommand, EntityCommands, RunSystemOnce as _},
+    prelude::*,
+    ui::Val::*,
+};
+
+use super::{interaction::InteractionPalette, palette::*};
+use crate::spawn::SpawnExt;
+
+/// An extension trait to improve the ergonomics of spawning UI widgets.
+pub trait Widgets {
+    /// Spawn a [`Button`].
+    fn button(&mut self, text: impl Into<String>) -> EntityCommands;
+
+    /// Spawn a [`Header`].
+    fn header(&mut self, text: impl Into<String>) -> EntityCommands;
+
+    /// Spawn a [`Label`].
+    fn label(&mut self, text: impl Into<String>) -> EntityCommands;
+}
+
+impl<T: SpawnExt> Widgets for T {
+    fn button(&mut self, text: impl Into<String>) -> EntityCommands {
+        self.spawn_with(Button::new(text))
+    }
+
+    fn header(&mut self, text: impl Into<String>) -> EntityCommands {
+        self.spawn_with(Header::new(text))
+    }
+
+    fn label(&mut self, text: impl Into<String>) -> EntityCommands {
+        self.spawn_with(Label::new(text))
+    }
+}
+
+/// A simple button with text.
+pub struct Button {
```

Comment:
Maybe rename this to `ButtonWidget` to avoid naming conflicts with `Bevy`.
Also, does this need to be `pub`?

Reply:
The naming conflict shouldn't be an issue since the intended use is `widget::Button`. Unless it's considered that it's an issue regardless. In my own template I have this as `widget::MenuButton` to avoid the conflict.

It needs to be `pub` because you might not be using the `Widgets` extension trait, which only provides spawning functionality, not adding functionality (which you could use to "inherit" from `Button`).